### PR TITLE
Floc rewrite to include entrance tank

### PR DESCRIPTION
**Pull request by monroews**

### CODE REVIEW

File: aguaclara/design/floc.py [hunk 1]
```diff
@@ -1,307 +1,272 @@
...
+   """
+
+   # Increased both to provide a safety margin on flocculator head loss and
+   # to simultaneously scale back on the actual collision potential we are
+   # trying to achieve.
+   # Originally calculated to be 2.3 from the equations:
+
+   # K_MINOR_FLOC_BAFFLE = (1/VC_BAFFLE_RATIO - 1)**2
+   BAFFLE_K = 2.5
+   CHANNEL_N_MIN = 2
```
Comment (eak24) [inline aguaclara/design/floc.py]: These class constants should go into the constructor. If we leave them as class vars, they are impossible to change (or harder to change) in the client code. Since they are optional params, I don't think that would overburden any callers.

**Comment by monroews** [comment]
The tests need to be updated to match the results given by this code. The original code didn't include the effect of having an entrance tank in the flocculator. I recommend creating the tests to match these results and then I will have students in CEE 4520 do a more rigorous test of the flocculator design to make sure that it works or to find where it fails.

File: aguaclara/design/floc.py [hunk 2]
```diff
@@ -1,307 +1,272 @@
...
+   HS_RATIO_MIN = 3
+   RATIO_MAX_HS = 6
+   SDR = 41
+
+
+   def __init__(self, Q=20 * u.L/u.s, temp=25 * u.degC,
```
Comment (eak24) [inline aguaclara/design/floc.py]: Hmm, how should we separate the inputs in the constructor into their own 'concerns' for easy discovery? Maybe this is a non-issues... numpy has a multitude of keyword args in their function signatures... https://docs.scipy.org/doc/numpy/reference/generated/numpy.linspace.html It just means our documentation needs to describe all the inputs, and we should have some automated checking to ensure each described function has all parameters fully (type and name) described. [I added another issue for that.](https://github.com/AguaClara/aguaclara/issues/162) This looks good for now, as long as the class constants are added I think!

File: aguaclara/design/floc.py [hunk 3]
```diff
@@ -1,307 +1,272 @@
...
+       possible channel width and the maximum length of the channels.
+       Round up to the next even number (factor of 2 shows up twice in equation)
+       The channel width must be greater than the hydraulic width that ensure baffle overlap.
+       Based on the equation for the flocculator volume
+       volume = ([max_L*channel_n] - entrancetank_L)*max_W * downstream_H
+       :returns: number of channels
+       :rtype: float * dimensionless
+       """
+       min_hydraulic_W = np.amax(np.array([1,(self.max_W/self.W_min_HS_ratio).to(u.dimensionless)])) * self.W_min_HS_ratio
+       return 2*np.ceil(((self.vol/(min_hydraulic_W * self.downstream_H) + self.entrancetank_L)/(2*self.max_L)).to(u.dimensionless))
```
Comment (eak24) [inline aguaclara/design/floc.py]: I wonder if separating this equation into smaller equations with vars that have descriptive names could make this more reasonable. As it stands, this is hard to parse and read through. For instance: 

```
ratio_of_widths = (self.max_W/self.W_min_HS_ratio).to(u.dimensionless)
# 1 is the lower bound because the width can't get smaller than the min
optimal_width_ratio = np.amax(np.array([1,ratio_of_widths]))
min_hydraulic_W = optimal_width_ratio * self.W_min_HS_ratio
length_without_headloss_volume = self.vol/(min_hydraulic_W * self.downstream_H) + self.entrancetank_L
ideal_number_of_channels = (length_without_headloss_volume)/(self.max_L)
actual_even_number_of_channels = 2*np.ceil((ideal_number_of_channels/2).to(u.dimensionless))
return actual_even_number_of_channels
```

Comment (eak24) [reply aguaclara/design/floc.py]: The idea is that the variable names provide the documentation. This eliminates the need for comments that often get way out of date. And since we're within the scope of the function, you can use nice long and descriptive var names, until it almost reads like english. That is the goal of all code: to be as readable as possible

File: aguaclara/design/floc.py [hunk 4]
```diff
@@ -1,307 +1,272 @@
...
+       - wider than hydraulic requirement to meet H/S ratio
+       Create a dimensionless array of the 3 requirements and then get the maximum
+
+       :returns: Channel width
+       :rtype: float * meter
+       """
+       channel_est_W = (self.vol / (self.downstream_H * (self.channel_n * self.max_L - self.entrancetank_L))).to(u.m)
+       # The channel may need to wider than the width that would get the exact required volume.
+       # In that case we will need to shorten the flocculator
+       channel_W = np.amax(np.array([1,(ha.HUMAN_W_MIN/channel_est_W).to(u.dimensionless),(self.W_min_HS_ratio/channel_est_W).to(u.dimensionless)])) * channel_est_W
```
Comment (eak24) [inline aguaclara/design/floc.py]: break up and simplify as seen above

File: aguaclara/design/floc.py [hunk 5]
```diff
@@ -1,307 +1,272 @@
...
+       volume will be greater than necessary. Bring the volume back to the design
+       volume by shortening the flocculator in this case. This design approach
+       will produce flocculators that are the same length as the max_L that was
+       specified in many cases. The flocculator will be less than the specified
+       length especially for cases with only one or two sed tanks.
+       channel_L = (vol/(channel_W * downstream_H) + entrancetank_L)/channel_n
+       :returns: Channel length
+       :rtype: float * meter
+       """
+       channel_L = ((self.vol / (self.channel_W * self.downstream_H) + self.entrancetank_L)/self.channel_n).to(u.m)
```
Comment (eak24) [inline aguaclara/design/floc.py]: break up and simplify as seen above

File: aguaclara/design/floc.py [hunk 6]
```diff
@@ -1,307 +1,272 @@
...
+       """"Return the maximum distance between expansions for the largest
+       allowable H/S ratio.
+       :returns: Maximum expansion distance
+       :rtype: float * meter
+       Examples
+       --------
+       exp_dist_max(20*u.L/u.s, 40*u.cm, 37000, 25*u.degC, 2*u.m)
+       0.375 meter
+       """
+       return (((self.BAFFLE_K / (2 * pc.viscosity_kinematic(self.temp) * (self.vel_grad_avg ** 2))) *
```
Comment (eak24) [inline aguaclara/design/floc.py]: break up and simplify as seen above

File: aguaclara/design/floc.py [hunk 7]
```diff
@@ -1,307 +1,272 @@
...
+       return (self.downstream_H / self.expansion_n).to(u.cm)
+
+   @property
+   def baffle_S(self):
+       """Return the spacing between baffles.
+       :returns: Spacing between baffles
+       :rtype: int
+       """
+
+       return ((self.BAFFLE_K /
```
Comment (eak24) [inline aguaclara/design/floc.py]: break up and simplify as seen above

File: aguaclara/design/floc.py [hunk 8]
```diff
@@ -1,307 +1,272 @@
...
+       of the entrance tank. Need to review the design to see if this is a good
+       assumption.
+       D_{Pipe} = \sqrt{ \frac{8 A_{Tank}}{\pi t_{Drain}} \sqrt{ \frac{h_0 \sum K}{2g} } }
+       :returns: list of designed values
+       :rtype: float * centimeter
+       """
+
+       drain_K = minorloss.PIPE_ENTRANCE_K_MINOR + minorloss.PIPE_ENTRANCE_K_MINOR +minorloss.PIPE_EXIT_K_MINOR
+       tank_A = 2* self.channel_L * self.channel_W
+       drain_D = (np.sqrt(8 * tank_A/(np.pi* self.drain_t) * np.sqrt(self.downstream_H * drain_K/(2*u.standard_gravity)))).to_base_units()
```
Comment (eak24) [inline aguaclara/design/floc.py]: A little better, but I think this computation is still a bit too long. Maybe this should actually be using a function defined elsewhere?

**Review by eak24 — CHANGES_REQUESTED** [review]
Looks good, my main comment is that it needs to be more readable. The equations are too long. I broke up one into what I think were OK names? It was very hard to reason through before breaking it up, and took me quite a while (8 min?) and that was just a very simple equation!

**Comment by monroews** [comment]
@oliver-leung can you 

- make the changes recommended by Ethan to make this code easier to read, 
- fix the tests
- create a new release of aguaclara with this code
And do this all within the next few days so that I can create an assignment using this code for CEE 4520?

**Comment by oliver-leung** [comment]
I'm sorry, I haven't had the time to work on this, and I have a lot of assignments in my classes to catch up on. I'll try to finish it in the next two days!

**Comment by oliver-leung** [comment]
I ran the tests, and they're running with different values. I want to assume that the new values are correct, but it's calculating the retention time as 3 seconds now (definitely not what we want!). Could you recheck the calculations for `floc.retention_time`?

Also, some functions are no longer there, like `floc.w_min` - are they not needed for the design logic?

**Comment by monroews** [comment]
Change the default Gt value to 37000. It is currently incorrectly set to 3700. That will change the residence time to be reasonable.
I have deleted functions that are no longer needed.

**Comment by monroews** [comment]
Should I squash and merge?
I'm wondering if we could change our tests to just check the final design output. That should check all of the code. Then we wouldn't need to check intermediate calculations.

**Comment by oliver-leung** [comment]
Apologies, I thought that Ethan's readability changes were implemented, but they haven't been yet. If you're okay with them being worked on later, you may squash and merge, and we can clean things up later!

**Comment by oliver-leung** [comment]
@monroews These changes are now on v0.0.20, so you can upgrade the package to get the new functions!
